devtools/x/src/cargo.rs

```diff
@@ -437,3 +392,153 @@ impl<'a> CargoCommand<'a> {
         }
     }
 }
+
+/// Arguments for the Cargo package selector.
+#[derive(Debug, StructOpt)]
+pub struct CargoPackageArgs {
+    #[structopt(long, short, number_of_values = 1)]
+    /// Run on the provided packages
+    pub(crate) package: Vec<String>,
+    #[structopt(long, short)]
+    /// Run on packages changed since the merge base of this commit
+    changed_since: Option<String>,
+    #[structopt(long)]
+    /// Run on all packages in the workspace
+    pub(crate) workspace: bool,
+}
+
+impl CargoPackageArgs {
+    pub fn to_cargo_packages<'a>(&'a self, xctx: &'a XContext) -> Result<CargoPackages<'a>> {
+        // Mutually exclusive options -- only one of these can be provided.
+        {
+            let mut exclusive = vec![];
+            if self.changed_since.is_some() {
+                exclusive.push("--changed-since");
+            }
+            if !self.package.is_empty() {
+                exclusive.push("--package");
+            }
+            if self.workspace {
+                exclusive.push("--workspace");
+            }
+
+            if exclusive.len() > 1 {
+                let err_msg = exclusive.join(", ");
+                return Err(anyhow!("can only specify one of {}", err_msg));
+            }
+        }
+
+        if self.workspace {
+            Ok(CargoPackages::workspace())
+        } else if !self.package.is_empty() {
+            Ok(CargoPackages::includes(
+                self.package.iter().map(|s| s.as_str()),
+            ))
+        } else if let Some(base) = &self.changed_since {
+            let affected_set = changed_since_impl(&xctx, &base)?;
+
+            Ok(CargoPackages::includes(
+                affected_set
+                    .packages(DependencyDirection::Forward)
+                    .map(|package| package.name()),
+            ))
+        } else {
+            CargoPackages::default_cwd(xctx)
+        }
+    }
+}
+
+/// Package selector for Cargo commands.
+#[derive(Clone, Debug)]
+pub struct CargoPackages<'a> {
+    includes: CargoPackageInclude<'a>,
+    excludes: BTreeSet<&'a str>,
+}
+
+impl<'a> CargoPackages<'a> {
+    /// Returns a new `CargoPackages` that selects all packages in this workspace.
+    pub fn workspace() -> Self {
+        Self {
+            includes: CargoPackageInclude::Workspace,
+            excludes: BTreeSet::new(),
+        }
+    }
+
+    /// Returns a new `CargoPackages` that selects the default set of packages for the current
+    /// working directory. This may either be the entire workspace or a local package.
+    pub fn default_cwd(xctx: &XContext) -> Result<Self> {
+        let includes = if project_is_root(&xctx.config().cargo_config())? {
+            CargoPackageInclude::Workspace
```

Comment:
Hmm, little funky naming... took a while to make it clear how you were returning the full workspace when reading CargoPackages.    Move some of the documentation up to the type?

Reply:
TODO: document design requirements and decisions here.

Reply:
Done.